scipy/linalg/_decomp_cossin.py

```diff
@@ -0,0 +1,186 @@
+        of subblocks X11,X12,X21,X22, when ``p=q=None``
+    partitioning : (int, int) or None
+        shape of the upper left block X11, if omitted, X is assumed to be given
+        in iterable form
+
+    Returns
+    -------
+    U : (m, m) ndarray
+        The block diagonal unitary or orthogonal matrix consisting of the
+        blocks U1 (p-by-p) and U2 (m-p)-by-(m-p)
+    CS : (m, m) ndarray
+        The cosine-sine factor with the structure described above
+    VDH : (m, m) ndarray
+        The block diagonal unitary or orthogonal matrix consisting of the
+        blocks V1 (q-by-q) and V2 (m-q)-by-(m-q)
+
+    Examples
+    --------
+    >>> from scipy.linalg import cossin
+    >>> from scipy.stats import unitary_group
+    >>> x = unitary_group.rvs(4)
+    >>> u, cs, vdh = cossin(x, (2,2))
+    >>> np.allclose(x, u @ cs @ vdh)
+    True
+
+    References
+    ----------
+    .. [1] : Brian D. Sutton. Computing the complete CS decomposition. Numer.
+           Algorithms, 50(1):33-65, 2009.
+
+    """
+    if partitioning:
+        X = _asarray_validated(X, check_finite=True)
+        if not np.equal(*X.shape):
+            raise ValueError("Cosine Sine decomposition only supports square"
+                             " matrices, got {}".format(X.shape))
+        p, q = partitioning
+        return _cossin(x11=X[:p, :q], x12=X[:p, q:], x21=X[p:, :q],
+                       x22=X[p:, q:])
+
+    if not isinstance(X, Iterable):
+        raise ValueError("When top_left_shape=None, X must be Iterable")
+
+    X = list(X)
+    if len(X) != 4:
+        raise ValueError("When partitioning=None, "
+                         "exactly four submatrices should be in X, got {}"
+                         .format(len(X)))
+
+    x11, x12, x21, x22 = X
+
+    p, q = np.shape(np.atleast_2d(x11))
+    mmp, mmq = np.shape(np.atleast_2d(x22))
+
+    p2, mmq2 = np.shape(np.atleast_2d(x12))
+    mmp2, q2 = np.shape(np.atleast_2d(x21))
+
+    if p != p2 or q != q2 or mmp != mmp2 or mmq != mmq2 or mmq + q != mmp + p:
+        shapes = [np.shape(np.atleast_2d(subx)) for subx in X]
+        raise ValueError("invalid submatrix dimensions {} to build a square "
+                         "matrix, maybe missing partitioning?".format(shapes))
+    return _cossin(x11, x12, x21, x22)
+
+
+def _cossin(x11, x12, x21, x22):
+    p, q = np.shape(x11)
+    m = np.shape(x22)[0] + p
+
+    iscomp = np.iscomplexobj(x11)
+    func_name = "uncsd" if iscomp else "orcsd"
+
+    csd, csd_lwork = get_lapack_funcs([func_name, func_name + "_lwork"],
+                                      [x11])
+    lwork = _compute_lwork(csd_lwork, m=m, p=p, q=q)
+
+    lwork_args = (dict(lwork=lwork[0], lrwork=lwork[1]) if iscomp else
+                  dict(lwork=lwork))
+
+    csd_args = dict(x11=x11,
+                    x12=x12,
+                    x21=x21,
+                    x22=x22,
+                    compute_u1=True,
```

Comment:
They are always on?

Reply:
yeah, I thought turning off the computation could be considered an "advanced" flow, and those who need it could drop down to the LAPACK level. I have no use case in mind for skipping the computations, but if there are (the mere existence of these flags suggest it) they could save performance, I assume. So I'm okay to plumb `compute_u` and `compute_vh` back to the top.